### DISCUSSION TITLE
Derive a rust struct from CDDL

### DISCUSSION THREAD

**Discussion by TomzBench** [anweiss/cddl#125] [Ideas]
If I have some CDDL, is it possible to automatically instantiate a struct with fields using procedure macros?

IE:

```rust
#[cddl(path = "path/to/foo.cddl")]
struct Foo {}
```

The struct is "empty" but after the macro runs, the fields should be populated with rust representation of the CDDL description.  Then you can use this in combination with Serde Serialize/Deserialize macros and get encoder/decoder for your struct.

IE:
```rust
#[derive(Serialize,Deserialize)]
#[cddl(path = "path/to/foo.cddl")]
struct Foo{}
```

To deal with nesting structs, the proc macro attribute could be at the module level, and just generate all the structs from the cddl. If one is familiar with proc macros, but not familiar with this crate, do you have any recommendations on how to go about implementing this? See any oversights?

**Comment by anweiss** [comment]
Hey @TomzBench, this is actually something I've been thinking about supporting for quite some time. I'm not as well-versed with proc macro implementations so feel free to give it a whirl and open a PR. As a starting point, you may want to look through the basic visitor trait in https://github.com/anweiss/cddl/blob/main/src/visitor.rs. You could probably implement the trait with a proc macro.

**Comment by SebastienGllmt** [comment]
This sounds more like something that would fit well within https://github.com/dcSpark/cddl-codegen

**Comment by anweiss** [comment]
First pass at this implemented in https://github.com/anweiss/cddl/pull/447
